### PR TITLE
Add Uniswap v2

### PR DESCRIPTION
**Pull request by this-username-is-taken**

### CODE REVIEW

File: subgraphs/uniswap-v2/schema.graphql [hunk 1]
```diff
@@ -0,0 +1,464 @@
+enum Network {
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:1]: You should update this with the latest shared schema in the repo root

File: subgraphs/uniswap-v2/schema.graphql [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+  amountOutUSD: BigDecimal!
+  ##### DexAmm-Specific #####
+  " The pool involving this transaction "
+  pool: LiquidityPool!
+}
+
+#Added Entities 
+
+# An account is a unique Ethereum address
+type Account @entity {
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:411]: Use convention as documented here: https://github.com/messari/subgraphs/blob/master/docs/Schema.md#internal-entities

File: subgraphs/uniswap-v2/schema.graphql [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+  " Address of the account "
+  id: ID!
+}
+
+type DailyActiveAccount @entity {
+  " { Days since Unix epoch }-{ Address of the account } "
+  id: ID!
+}
+
+type Transaction @entity {
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:421]: Is this still needed?

File: subgraphs/uniswap-v2/schema.graphql [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+  # populated from the Burn event
+  sender: Bytes
+  to: Bytes
+ 
+  # mark uncomplete in ETH case
+  needsComplete: Boolean!
+}
+
+# stores for USD calculations
+type Bundle @entity {
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:457]: You should probably want to use a more descriptive name for this

File: subgraphs/uniswap-v2/schema.graphql [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+  blockNumber: BigInt!
+  timestamp: BigInt!
+  # This is not the reverse of Mint.transaction; it is only used to
+  # track incomplete mints (similar for burns and swaps)
+  mints: [Mint!]
+  burns: [Burn!]
+}
+
+type Mint @entity {
+  # transaction hash + "-" + index in mints Transaction array
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:432]: Use double quote comments instead. These will show up in the subgraph playground when users browse the schema

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  BIGDECIMAL_ZERO
+} from './common/constants'
+
+export function handleNewPair(event: PairCreated): void {
+
+  let protocol = DexAmmProtocol.load(FACTORY_ADDRESS)
+
+  if (protocol === null) {
+    protocol = new DexAmmProtocol(FACTORY_ADDRESS)
+    protocol.name = "Uniswap v3"
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:17]: v2?

File: subgraphs/uniswap-v2/src/common/constants.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import { Address, BigDecimal, BigInt } from '@graphprotocol/graph-ts';
+
+export const PROTOCOL_ID = "1337"
+
+export const NETWORK_ETHEREUM = "ETHEREUM"
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/common/constants.ts:5]: You should update the named constants using the `namespaced` format here: https://github.com/messari/subgraphs/blob/40d5822bf6dbdbcae185a7ccdd77b526c9dc5cbd/subgraphs/yearn-v2/src/common/constants.ts#L7

File: subgraphs/uniswap-v2/src/common/Price.ts [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+ * If neither is, return 0
+ */
+ export function getTrackedVolumeUSD(
+  tokenAmount0: BigDecimal,
+  token0: Token,
+  tokenAmount1: BigDecimal,
+  token1: Token,
+  pool: LiquidityPool
+): BigDecimal {
+  let bundle = Bundle.load('1')
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/common/Price.ts:131]: Use a named constant for the ID

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    // create new bundle
+    let bundle = new Bundle('1')
+    bundle.ethPrice = BIGDECIMAL_ZERO
+    bundle.save()
+  }  
+
+  // create the tokens
+  let token0 = getOrCreateToken(event.params.token0)
+  let token1 = getOrCreateToken(event.params.token1)
+  let rewardToken0 = getOrCreateRewardToken(event.params.token0, "DEPOSIT")
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:31]: Use named constant for "DEPOSIT" as shown here: https://github.com/messari/subgraphs/blob/40d5822bf6dbdbcae185a7ccdd77b526c9dc5cbd/subgraphs/yearn-v2/src/common/constants.ts#L38

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  let rewardToken1 = getOrCreateRewardToken(event.params.token0, "DEPOSIT")
+
+  let pool = new LiquidityPool(event.params.pair.toHexString())
+
+  pool.protocol = "ETHEREUM"
+  pool.inputTokens =  [token0.id, token1.id]
+  pool.outputToken = token0.symbol + "/" + token1.symbol + "Uniswap v2 LP"
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:38]: The `outputToken` ID should be the address of the LP token

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  pool.rewardTokens = [rewardToken0.id, rewardToken1.id]
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:39]: No reward token for Uniswap

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  pool.totalValueLockedUSD = BIGDECIMAL_ZERO
+  pool.totalVolumeUSD = BIGDECIMAL_ZERO
+  pool.inputTokenBalances = [BIGDECIMAL_ZERO, BIGDECIMAL_ZERO]
+  pool.outputTokenSupply = BIGDECIMAL_ZERO
+  pool.outputTokenPriceUSD = BIGDECIMAL_ZERO
+  pool.createdTimestamp = event.block.timestamp
+  pool.createdBlockNumber = event.block.number
+  pool.name = protocol.name + " " + token0.symbol + "/" + token1.symbol
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:47]: Should get this from the LP token

File: subgraphs/uniswap-v2/src/Factory.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  pool.symbol = token0.symbol + "/" + token1.symbol
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/Factory.ts:48]: Should get this from the LP token

File: subgraphs/uniswap-v2/src/common/Price.ts [hunk 14]
```diff
@@ -0,0 +1,183 @@
+/* eslint-disable prefer-const */
+import { BigDecimal, Address, BigInt } from '@graphprotocol/graph-ts/index'
+import { factoryContract, UNTRACKED_PAIRS } from './helpers'
+import { LiquidityPool, Token, Bundle } from '../../generated/schema'
+import { ZERO_ADDRESS, BIGDECIMAL_ZERO, BIGDECIMAL_ONE } from '../common/constants'
+
+const WETH_ADDRESS = '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'
+const USDC_WETH_PAIR = '0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc' // created 10008355
+const DAI_WETH_PAIR = '0xa478c2975ab1ea89e8196811f51a7b7ade33eb11' // created block 10042267
+const USDT_WETH_PAIR = '0x0d4a11d5eeaac28ec3f61d100daf4d40471f1852' // created block 10093341
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/src/common/Price.ts:10]: These should go into the `constants` file

File: subgraphs/uniswap-v2/subgraph.yaml [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+          handler: handleNewPair
+templates:
+  - kind: ethereum/contract
+    name: Pair
+    network: mainnet
+    source:
+      abi: Pair
+    mapping:
+      kind: ethereum/events
+      apiVersion: 0.0.5
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/subgraph.yaml:46]: upgrade 0.0.6

File: subgraphs/uniswap-v2/schema.graphql [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+  id: ID! # txn hash
+  blockNumber: BigInt!
+  timestamp: BigInt!
+  # This is not the reverse of Mint.transaction; it is only used to
+  # track incomplete mints (similar for burns and swaps)
+  mints: [Mint!]
+  burns: [Burn!]
+}
+
+type Mint @entity {
```
Comment (this-username-is-taken) [inline subgraphs/uniswap-v2/schema.graphql:431]: Why not use `Deposit` and `Withdraw` entities instead?

File: subgraphs/uniswap-v2/schema.graphql (hunk 3 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/schema.graphql:421]: The Transaction entity?

File: subgraphs/uniswap-v2/schema.graphql (hunk 1 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/schema.graphql:1]: Updated!

File: subgraphs/uniswap-v2/schema.graphql (hunk 2 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/schema.graphql:411]: Updated!

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 6 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:17]: Updated!

File: subgraphs/uniswap-v2/src/common/constants.ts (hunk 7 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/common/constants.ts:5]: Fixed

File: subgraphs/uniswap-v2/src/common/Price.ts (hunk 8 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/common/Price.ts:131]: Updated

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 9 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:31]: Done!

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 10 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:38]: fixed

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 11 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:39]: Removed

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 12 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:47]: Done

File: subgraphs/uniswap-v2/src/Factory.ts (hunk 13 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/Factory.ts:48]: Done

File: subgraphs/uniswap-v2/src/common/Price.ts (hunk 14 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/src/common/Price.ts:10]: Moved to constants

File: subgraphs/uniswap-v2/subgraph.yaml (hunk 15 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/subgraph.yaml:46]: updated

File: subgraphs/uniswap-v2/schema.graphql (hunk 16 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/schema.graphql:431]: Its used to track values across events

File: subgraphs/uniswap-v2/schema.graphql (hunk 4 above)
Comment (steegecs) [reply subgraphs/uniswap-v2/schema.graphql:457]: Fixed

**Comment by 0xbe1** [comment]
@this-username-is-taken  What's the relationship between this PR and https://github.com/messari/subgraphs/pull/7?

**Comment by this-username-is-taken** [comment]
This one is abandoned. Please refer to #7, which is the latest work in progress.

**Pull request closed without merging**